### ISSUE TITLE
[TF 2.1] Error when converting LSTM model to a frozen graph using convert_variables_to_constants_v2()

### ISSUE THREAD

**Issue by jakesabathia2** [tensorflow/tensorflow#36391]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Mac
- TensorFlow installed from (source or binary): pip install tensorflow
- TensorFlow version (use command below): 2.1
- Python version: 2.7

I tried to froze a LSTM model build with tf.keras by using the `convert_variables_to_constants_v2` function with the following code

```
import tensorflow as tf

def build_model(vocab_size, embedding_dim, rnn_units, batch_size):
  model = tf.keras.Sequential([
    tf.keras.layers.Embedding(vocab_size, embedding_dim,
                              batch_input_shape=[batch_size, embedding_dim]),
    tf.keras.layers.LSTM(rnn_units,
                        return_sequences=True,
                        stateful=False,
                        recurrent_activation='sigmoid',
                        recurrent_initializer='glorot_uniform'),
    tf.keras.layers.Dense(vocab_size)
  ])
  return model

embedding_dim = 100
units = 256
vocab_size = 300
batch_size = 32

model = build_model(vocab_size, embedding_dim, units, batch_size)
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy')

from tensorflow.python.keras.saving import saving_utils as _saving_utils
from tensorflow.python.framework import convert_to_constants as _convert_to_constants

tf.keras.backend.set_learning_phase(False)
func = _saving_utils.trace_model_call(model)
concrete_func = func.get_concrete_function()
frozen_func = _convert_to_constants.convert_variables_to_constants_v2(concrete_func)
```
produces
```
Cannot find the Placeholder op that is an input to the ReadVariableOp.
```

**Comment by gadagashwini** [comment]
Could replicate issue on colab with Tf 2.1 and Tf-nightly.
Please find the gist [here](https://colab.research.google.com/gist/gadagashwini/483fa045f51d0112ddbd2039781b1eca/untitled372.ipynb). Thanks!

**Comment by jakesabathia2** [comment]
Thanks!

**Comment by qlzh727** [comment]
Hi @jakesabathia2, the code snippet you have is using a lot of internal python function, which access a lot implementation details and we don't ensure any API guarantee for them.

Can you provide more context about why you need to access the frozen graph? Also please check the model saving doc here https://www.tensorflow.org/guide/saved_model.

**Comment by Sam-Persoon** [comment]
> Hi @jakesabathia2, the code snippet you have is using a lot of internal python function, which access a lot implementation details and we don't ensure any API guarantee for them.
> 
> Can you provide more context about why you need to access the frozen graph? Also please check the model saving doc here https://www.tensorflow.org/guide/saved_model.

I also came along this issue yesterday. I need the frozen graph to list all used Ops for security reasons (models that I did not create, could read or write from/to disk). As far as I know/found, this is only possible using the frozen graph. 

For example: I want to search for MatchingFiles, ReadFile or WriteFile ops in a model in SavedModel format. The library part of the saved model proto lists this, but it is documented that this is still experimental and might be removed.

**Comment by Sam-Persoon** [comment]
Any updates on this?

**Comment by qlzh727** [comment]
Adding Kathy for the save model question.

**Comment by k-w-w** [comment]
+@gargn  Should convert_variables_to_constants_v2 be working for this case?

**Comment by Sam-Persoon** [comment]
> +@gargn Should convert_variables_to_constants_v2 be working for this case?

Another way of achieving this, is also fine. But it should be possible at least as all the information is in the graph anyway.

**Comment by luvwinnie** [comment]
any progress on this issues?

**Comment by moniGra** [comment]
I was having this issue too with my keras model (TF 2.0 and nightly reported "Cannot find the Placeholder op that is an input to the ReadVariableOp." while trying to convert to SM model) - I managed to solve this by following the suggestions from here: https://stackoverflow.com/questions/51858203/cant-import-frozen-graph-with-batchnorm-layer/52823701#52823701 
Basically:
1) set trainable to false in each element of keras model
2) before converting (in my case to SM, then from SM to frozen graph):
`keras.backend.clear_session() 
tf.keras.backend.set_learning_phase(0)  `

**Comment by luvwinnie** [comment]
> I was having this issue too with my keras model (TF 2.0 and nightly reported "Cannot find the Placeholder op that is an input to the ReadVariableOp." while trying to convert to SM model) - I managed to solve this by following the suggestions from here: https://stackoverflow.com/questions/51858203/cant-import-frozen-graph-with-batchnorm-layer/52823701#52823701
> Basically:
> 
> 1. set trainable to false in each element of keras model
> 2. before converting (in my case to SM, then from SM to frozen graph):
>    `keras.backend.clear_session()    `
>    `tf.keras.backend.set_learning_phase(0)`

which version of the tensorflow did you use?
I have tried by setting learning phase to 0 but, it doesn't work.
I also tried with this link methods to freeze my graph, but also doesn't work, i think it is a problem with [this link](https://medium.com/@xianbao.qian/how-to-export-the-inference-model-in-tf2-279ac233915c) methods.

**Comment by luvwinnie** [comment]
Hmm. I look at the [test code](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/convert_to_constants_test.py) of convert_variables_to_constants_v2, It seems like when our model have the LSTM or GRU's layers, what we need to do is by pass one more argument `lower_control_flow=False` in order to make the convert_variables_to_constants_v2 work.

The below code is an example to freeze the model, hope it helps
```
full_model = tf.function(lambda x: model(x))
full_model = full_model.get_concrete_function(
    tf.TensorSpec(model.inputs[0].shape, model.inputs[0].dtype))

# Get frozen ConcreteFunction   
frozen_func = convert_variables_to_constants_v2(full_model,lower_control_flow=False)
frozen_func.graph.as_graph_def()

layers = [op.name for op in frozen_func.graph.get_operations()]
# Save frozen graph from frozen ConcreteFunction to hard drive
tf.io.write_graph(graph_or_graph_def=frozen_func.graph,
                  logdir="./frozen_models",
                  name="frozen_graph.pb",
                  as_text=False)
```

**Comment by xixiddd** [comment]
The current implementation of convert_variables_to_constants_v2 does not support graphs that contain embedding related ops, see https://github.com/tensorflow/tensorflow/blob/667485ac4aed337194b3c6d1bf1e2230b214537b/tensorflow/python/framework/convert_to_constants.py#L665
and you used tf.keras.layers.Embedding in your codes, is that why you got this error?

**Comment by Mistobaan** [comment]
I am seeing this too, thanks @xixiddd for the insights. That function is also used by the TensorRT examples to create optimized graphs.

**Comment by hatooku** [comment]
Also seeing this, I think my issue is related to embedding layers as @xixiddd mentioned. Is there any plan for adding support for embedding related ops?

**Comment by sachinprasadhs** [comment]
Did not face any issues while running your code on Tf Nightly 2.6.0-dev20210524, please find the gist [here](https://colab.research.google.com/gist/sachinprasadhs/3615ebc38df5e44facd934f33b3ef38a/35650.ipynb#scrollTo=kGg_oY4k3pqV).
